.github/workflows/docker_build_push.yml

```diff
@@ -9,39 +14,87 @@ on:
         type: string
 
 env:
-  REGISTRY: ghcr.io
   IMAGE_PATH: ghcr.io/${{ github.repository }}-${{ inputs.NOTEBOOK_TYPE }}
+  TARFILE_NAME: ${{ inputs.NOTEBOOK_TYPE }}-image.tar
 
 jobs:
-  build-and-push-image:
+  build:
     runs-on: ubuntu-latest
     permissions:
       contents: read
-      packages: write
 
     steps:
       - name: Checkout repository
-        uses: actions/checkout@v2
-
-      - name: Log in to the Container registry
-        uses: docker/login-action@v2.1.0
-        with:
-          registry: ${{ env.REGISTRY }}
-          username: ${{ github.actor }}
-          password: ${{ secrets.GITHUB_TOKEN }}
+        uses: actions/checkout@v4
 
       - name: Extract metadata (tags, labels) for Docker
         id: meta
         uses: docker/metadata-action@v4.1.1
         with:
           images: ${{ env.IMAGE_PATH }}
+          tags: |
+            type=ref,event=branch
+            type=ref,event=pr
+            type=semver,pattern={{version}} # major.minor.patch
+            type=semver,pattern={{major}}.{{minor}}
 
-      - name: Build and push Docker image
-        uses: docker/build-push-action@v3.2.0
+      - name: Build Docker image for scanning, but don't push to ghcr.io yet
+        uses: docker/build-push-action@v6.4.0
         with:
           context: .
           build-args: notebook_type=${{ inputs.NOTEBOOK_TYPE }}
-          push: true
+          push: false
+          outputs: type=tar,dest=${{ env.TARFILE_NAME }}
           tags: ${{ steps.meta.outputs.tags }}
           labels: ${{ steps.meta.outputs.labels }}
+
+      - name: Upload tarball for use by Trivy job
+        uses: actions/upload-artifact@v4
+        with:
+          name: ${{ env.TARFILE_NAME }}
+          path: ${{ env.TARFILE_NAME }}
+
+    outputs:
+      tags: ${{ steps.meta.outputs.tags }}
+      tarfile_artifact: ${{ env.TARFILE_NAME }}
+
+  trivy-scan:
+    needs: build
+    uses: "./.github/workflows/trivy.yml"
+    with:
+      NOTEBOOK_TYPE: ${{ inputs.NOTEBOOK_TYPE }}
+      SOURCE_TYPE: tar
```

Comment:
wondering why you decided to setup this scan is with type `tar` while the periodic scan is `image`?

Reply:
@zaro0508 An `image` scan is a scan on an image that has been pushed to the registry (the GitHub Container Registry, GHCR).  Here we perform the vulnerability scan __before__ we push the newly built image to the registry.  So we save it as a file archive (a `.tar` file) and have Trivy scan the archive.   If the scan passes, only then do we push the new image to GHCR.

The periodic scan is different:  Here we have already pushed the image to GHCR.  (We want to re-scan it in case new vulnerabilities have been discovered.) So we perform an "image scan".